File: src/Day08.java
import javautils.JavaUtils.JGrid;
import javautils.JavaUtils.JPoint2D;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import java.util.stream.StreamSupport;

import static java.util.function.Function.identity;
import static java.util.stream.Collectors.mapping;
import static java.util.stream.Collectors.toList;
import static java.util.stream.IntStream.range;
import static java.util.stream.Stream.iterate;

public class Day08 {

  public static void main(String[] args) throws IOException {
    if (args.length == 0) {
      List<String> testInput = Files.readAllLines(Path.of("src/08t.txt".trim()));
      assert part1(testInput) == 14 : "Expected 14 but got " + part1(testInput);

      List<String> input = Files.readAllLines(Path.of("src/08.txt".trim()));
      System.out.println(part1(input));
      assert part2(testInput) == 34 : "Expected 34 but got " + part2(testInput);
      System.out.println(part2(input));
    } else {
      List<String> input = Files.readAllLines(Path.of(args[0]));
      System.out.println(part1(input));
      System.out.println(part2(input));
    }
  }

  public static Map<Character, List<JPoint2D>> antennas(JGrid<Character> grid) {
    return StreamSupport.stream(grid.spliterator(), false)
        .filter(antenna -> grid.get(antenna) != '.')
        .collect(Collectors.groupingBy(grid::get, mapping(identity(), toList())));
  }

  public static int part1(List<String> input) {
    final var grid = JGrid.asCharGrid(input);
    return (int) antennas(grid).values().stream()
        .flatMap(list -> range(0, list.size()).boxed()
            .flatMap(i -> range(i + 1, list.size()).boxed()
                .flatMap(j -> {
                  JPoint2D first = list.get(i);
                  JPoint2D second = list.get(j);
                  return Stream.of(
                      second.times(2).minus(first),
                      first.times(2).minus(second)
                  );
                })
            ))
        .distinct()
        .filter(grid::contains)
        .count();
  }

  public static int part2(List<String> input) {
    final var grid = JGrid.asCharGrid(input);
    return (int) antennas(grid).values().stream()
        .flatMap(list ->
            range(0, list.size()).boxed()
                .flatMap(i -> range(i + 1, list.size())
                    .boxed().flatMap(j -> {
                      JPoint2D first = list.get(i);
                      JPoint2D second = list.get(j);
                      JPoint2D diff = second.minus(first);
                      var sequence1 = iterate(second.plus(diff), grid::contains, p -> p.plus(diff));
                      var sequence2 = iterate(first.minus(diff), grid::contains, p -> p.minus(diff));
                      return concat(sequence1, sequence2, Stream.of(first, second));
                    })
                ))
        .distinct()
        .count();
  }

  @SafeVarargs
  private static <T> Stream<T> concat(Stream<T>... streams) {
    return Stream.of(streams)
        .reduce(Stream::concat)
        .orElseGet(Stream::empty);
  }

}

File: src/Day08.kt
import kotlin.io.path.Path
import kotlin.io.path.readText


fun main(args: Array<String>) {


    fun antennas(grid: Grid<Char>): Map<Char, List<Point2D>> {
        return grid
            .asSequence()
            .filter { (_, c) -> c != '.' }
            .groupBy { it.second }
            .mapValues { it.value.map { it.first } }
    }


    fun Grid<Char>.findAntiNodes(first: Point2D, second: Point2D): Sequence<Point2D> {
        val diff = second - first
        val sequence1 = generateSequence(second + diff) { it + diff }.takeWhile { it in this }
        val sequence2 = generateSequence(first - diff) { it - diff }.takeWhile { it in this }
        return sequence1 + sequence2 + first + second
    }

    fun part1Komprehensions(input: List<String>): Int {
        return doFlatMapIterable(
            { listOf(input.toGrid()) },
            { grid -> antennas(grid).values },
            { _, antennas -> antennas.withIndex() },
            { _, antennas, (i, _) -> (i + 1 until antennas.size).asIterable() },
            { grid, antennas, (_, antenna1), j ->
                val antenna2 = antennas[j]
                sequenceOf(antenna2 * 2 - antenna1, antenna1 * 2 - antenna2).filter { it in grid }.asIterable()
            })
            .distinct()
            .count()
    }

    fun part2Komprehensions(input: List<String>) = doFlatMapIterable(
        { listOf(input.toGrid()) },
        { grid -> antennas(grid).values },
        { _, antennas -> antennas.withIndex() },
        { _, antennas, (i, _) -> (i + 1 until antennas.size).asIterable() },
        { grid, antennas, (_, antenna), j -> grid.findAntiNodes(antenna, antennas[j]).asIterable() })
        .distinct()
        .count()

    fun part1(input: List<String>): Int {
        val grid = input.toGrid()
        return antennas(grid).values.flatMap {
            it.flatMapIndexed { i, first ->
                (i + 1 until it.size).flatMap { j ->
                    val second = it[j]
                    listOf(second * 2 - first, first * 2 - second)
                }
            }
        }
            .distinct()
            .count { it in grid }
    }

    fun part2(input: List<String>): Int {
        val grid = input.toGrid()
        return antennas(grid).values.flatMap {
            it.flatMapIndexed { i, curPoint ->
                (i + 1 until it.size).flatMap { j -> grid.findAntiNodes(curPoint, it[j]) }
            }
        }
            .distinct()
            .size
    }

    if (args.isEmpty()) {
        val testInput = readInput("08t")
        check(part1(testInput) == 14) { "Expected 14 but got ${part1(testInput)}" }

        val input = readInput("08")
        part1(input).println()
        part1Komprehensions(input).println()
        check(part2(testInput) == 34) { "Expected 34 but got ${part2(testInput)}" }
        part2(input).println()
        part2Komprehensions(input).println()
    } else {
        val input = Path(args[0]).readText().trim().lines()
        part1(input).println()
        part2(input).println()
        part1Komprehensions(input).println()
        part2(input).println()
    }
}


